test/acceptance/tests/hcp/hcp_test.go

```diff
@@ -155,6 +150,402 @@ func TestHCP(t *testing.T) {
+		TerraformDir: tfDir,
+		NoColor:      true,
+		Logger:       terratestLogger.Discard,
+	})
+	return terraformOptions, outputs
+}
+
+func terraformDestroy(t *testing.T, tfOpts *terraform.Options, noCleanupOnFailure bool) {
+	if noCleanupOnFailure && t.Failed() {
+		logger.Log(t, "skipping resource cleanup because -no-cleanup-on-failure=true")
+	} else {
+		terraform.Destroy(t, tfOpts)
+	}
+}
+
+func expectCurlOutput(t *testing.T, task *helpers.MeshTask, expected string) {
+	retry.RunWith(timeouts["mesh-task"], t, func(r *retry.R) {
+		curlOut, err := task.ExecuteCommand("basic", `/bin/sh -c "curl localhost:1234"`)
+		r.Check(err)
+		if !strings.Contains(curlOut, expected) {
+			r.Errorf("response was unexpected: %q", curlOut)
+		}
+	})
+}
+
+func upsertIntention(t *testing.T, consulClient *api.Client, action api.IntentionAction, src, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, _, err := consulClient.ConfigEntries().Set(&api.ServiceIntentionsConfigEntry{
+			Kind:      api.ServiceIntentions,
+			Name:      dst.Name,
+			Partition: dst.Partition,
+			Namespace: dst.Namespace,
+			Sources: []*api.SourceIntention{
+				&api.SourceIntention{
+					Name:      src.Name,
+					Partition: src.Partition,
+					Namespace: src.Namespace,
+					Action:    action,
+				},
+			},
+		}, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func deleteIntention(t *testing.T, consulClient *api.Client, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, err := consulClient.ConfigEntries().Delete(api.ServiceIntentions, dst.Name, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func upsertExportedServices(t *testing.T, consulClient *api.Client, src, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, _, err := consulClient.ConfigEntries().Set(&api.ExportedServicesConfigEntry{
+			Name:      dst.Partition,
+			Partition: dst.Partition,
+			Services: []api.ExportedService{{
+				Name:      dst.Name,
+				Namespace: dst.Namespace,
+				Consumers: []api.ServiceConsumer{{Partition: src.Partition}},
+			}},
+		}, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func deleteExportedServices(t *testing.T, consulClient *api.Client, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, err := consulClient.ConfigEntries().Delete(api.ExportedServices, dst.Partition, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func consulClient(t *testing.T, addr, token string) (*api.Client, ConsulState, error) {
+	cfg := api.DefaultConfig()
+	cfg.Address = addr
+	cfg.Token = token
+	client, err := api.NewClient(cfg)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to create Consul client: %w", err)
+	}
+	state, err := recordConsulState(t, client)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to establish Consul state: %w", err)
+	}
+	return client, state, err
+}
+
+type ConsulState struct {
```

Comment:
I'm just curious, how does this deal with restoring the state of services?

Reply:
Currently it only handles ACL state that may be changed by the ACL controller. Do you think it is worth changing the name of the struct to something like `ConsulACLState`?

Reply:
I've left this as is for now. I've create a task in the backlog to revisit the acceptance test performance.